code/modules/minimap/minimap_ui.dm

```diff
@@ -0,0 +1,201 @@
+			placable_marker_states.Add(icon_state)
+			var/icon/marker_icon = icon('icons/obj/minimap/minimap_markers.dmi', icon_state)
+			placable_marker_images[icon_state] = icon2base64(marker_icon)
+
+		. += list(
+			"placable_marker_states" = placable_marker_states,
+			"placable_marker_images" = placable_marker_images,
+			"x" = 1,
+			"y" = 1
+		)
+
+	ui_data(mob/user)
+		..()
+		minimap_markers_list = list()
+		for (var/atom/target in src.minimap_datum.minimap_markers)
+			var/datum/minimap_marker/marker = src.minimap_datum.minimap_markers[target]
+			if (!marker.on_minimap_z_level)
+				continue
+
+			minimap_markers_list.Add(list(list(
+				"name" = marker.name,
+				"pos" = "[target.x], [target.y]",
+				"visible" = marker.visible,
+				"can_be_deleted" = marker.can_be_deleted_by_player,
+				"marker" = marker,
+				"index" = length(minimap_markers_list) + 1
+			)))
+
+		. = list(
+			"markers_visible" = markers_visible,
+			"selecting_coordinates" = src.minimap_controller.selecting_coordinates,
+			"minimap_markers" = minimap_markers_list
+		)
+
+		if (src.minimap_controller.selected_x && src.minimap_controller.selected_y)
+			. += list(
+				"x" = src.minimap_controller.selected_x,
+				"y" = src.minimap_controller.selected_y
+			)
+			src.minimap_controller.selected_x = null
+			src.minimap_controller.selected_y = null
+
+	ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+		var/mob/user
+		if(ismob(usr))
+			user = usr
+
+		if(is_incapacitated(user) || !(src.loc in user.equipped_list()))
+			return
+
+		switch (action)
+			if ("reset_scale")
+				if (istype(src.minimap_datum, /datum/minimap/z_level))
+					var/datum/minimap/z_level/minimap = src.minimap_datum
+					minimap.find_focal_point()
+
+			if ("toggle_visibility_all")
+				for (var/atom/target in src.minimap_datum.minimap_markers)
+					if (target.z != minimap_datum.z_level)
+						continue
+
+					var/datum/minimap_marker/marker = src.minimap_datum.minimap_markers[target]
+					if (markers_visible == TRUE)
+						marker.marker.alpha = 0
+						marker.visible = FALSE
+					else
+						marker.marker.alpha = 255
+						marker.visible = TRUE
+
+				if (markers_visible == TRUE)
+					markers_visible = FALSE
+				else
+					markers_visible = TRUE
+
+			if ("toggle_visibility")
+				var/list/list_entry = src.minimap_markers_list[params["index"]]
+				var/datum/minimap_marker/marker = list_entry["marker"]
+				if (marker.marker.alpha == 255)
+					marker.marker.alpha = 0
+					marker.visible = FALSE
+				else
+					marker.marker.alpha = 255
+					marker.visible = TRUE
+
+			if ("location_from_minimap")
+				minimap_controller.selecting_coordinates = TRUE
+
+			if ("new_marker")
+				var/name = params["name"]
+				var/icon_state = params["icon"]
+				var/x = params["pos_x"]
+				var/y = params["pos_y"]
+
+				if (!name || !icon_state || !x || !y)
+					message_admins("[name], [icon_state], [x], [y]")
```

Comment:
```suggestion
					message_admins("Invalid minimap marker: [name], [icon_state], [x], [y]")
```
or just remove this if it was only intended for local debugging

Reply:
probably worthwhile to log all markers somewhere, since it takes a name?